test/parsing.test.ts

```diff
@@ -745,6 +823,16 @@ describe('Zod Parsing', () => {
       //@ts-ignore
       assert.ok(typeof schema.partial === 'undefined');
     });
+
+    it('should return default value if parsing undefined', () => {
+      const schema = z.object({ a: z.string(), b: z.string() }).default({ a: 'hello', b: 'world' });
+      assert.deepEqual(schema.parse(undefined), { a: 'hello', b: 'world' });
+    });
+
+    it('should return default value if parsing undefined - func', () => {
+      const schema = z.object({ a: z.string(), b: z.string() }).default(() => ({ a: 'hello', b: 'world' }));
+      assert.deepEqual(schema.parse(undefined), { a: 'hello', b: 'world' });
+    });
```

Comment:
Can we add a test for nested values defaults? -- the following one is passing:

```ts
it('should return default value for nested values', () => {
    const schema = z.object({ a: z.string(), b: z.string().default("world") });
    assert.deepEqual(schema.parse({ a: "hello" }), { a: 'hello', b: 'world' });
});
```

Reply:
this is an extremely important test that needs to be written. thanks for flagging it

Reply:
I also wanted to create the following test case:

```ts
it('should return default value for nested values with default root', () => {
  const schema = z
    .object({
      a: z.string(),
      b: z.string().default(() => 'world').optional(),
    })
    .default({ a: 'hello' });
  assert.deepEqual(schema.parse(undefined), { a: 'hello', b: 'world' });
});
```

This does not work: the result is `{a: "hello", b: undefined}` instead of the expected value ; I am not yet sure why.

Also, as `default()` does not exist on `OptionalType`, `z.string().default('world').optional()` works while `z.string().optional().default('world')` fails type checking. Two things here:

– I think setting a default should imply `optional`, which is the whole point of having a default value in the first place.

– `default` should work with optional values when the underlying type is `Defaultable`.